File: chatuikit/src/main/java/com/cometchat/chatuikit/groupmembers/scopechange/ScopeAdapter.java
package com.cometchat.chatuikit.groupmembers.scopechange;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;
import androidx.appcompat.widget.AppCompatRadioButton;
import androidx.recyclerview.widget.RecyclerView;

import com.cometchat.chat.constants.CometChatConstants;
import com.cometchat.chat.models.Group;
import com.cometchat.chat.models.GroupMember;
import com.cometchat.chatuikit.R;

import java.util.List;

public class ScopeAdapter extends RecyclerView.Adapter<ScopeAdapter.RoleViewHolder> {
    private static final String TAG = ScopeAdapter.class.getSimpleName();
    private final Context context;
    private List<String> roles;
    private int selectedPosition = -1; // To keep track of the selected position
    private int initialPosition = -1;
    private Group group;

    private @ColorInt int itemTextColor;
    private @ColorInt int disableItemTextColor;
    private @StyleRes int itemTextAppearance;
    private @ColorInt int disableRadioButtonTint;
    private @ColorInt int radioButtonTint;
    private OnScopeSelectionChangeListener onScopeSelectionChangeListener;

    public ScopeAdapter(Context context, List<String> roles) {
        this.context = context;
        this.roles = roles;
    }

    @NonNull
    @Override
    public RoleViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.cometchat_list_scope_change_item_radio, parent, false);
        return new RoleViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RoleViewHolder holder, @SuppressLint("RecyclerView") int position) {
        String role = roles.get(position);
        holder.textViewRole.setText(role);

        holder.radioButton.setChecked(holder.getAbsoluteAdapterPosition() == selectedPosition);
        holder.textViewRole.setTextAppearance(itemTextAppearance);
        if ((CometChatConstants.SCOPE_MODERATOR.equalsIgnoreCase(group.getScope()) && role.equalsIgnoreCase(CometChatConstants.SCOPE_ADMIN))) {
            holder.radioButton.setEnabled(false);
            holder.radioButton.setButtonTintList(ColorStateList.valueOf(disableRadioButtonTint));
            holder.textViewRole.setTextColor(disableItemTextColor);
        } else {
            holder.textViewRole.setTextColor(itemTextColor);
            holder.radioButton.setButtonTintList(ColorStateList.valueOf(radioButtonTint));
            holder.radioButton.setEnabled(true);
            // Handle the item click
            holder.itemView.setOnClickListener(v -> {
                selectedPosition = holder.getAbsoluteAdapterPosition();
                notifyDataSetChanged();
                notifySelectionChangeListener();
            });

            // Handle the RadioButton click
            holder.radioButton.setOnClickListener(v -> {
                selectedPosition = holder.getAbsoluteAdapterPosition();
                notifyDataSetChanged();
                notifySelectionChangeListener();
            });
        }
    }

    public void setItemTextColor(@ColorInt int itemTextColor) {
        this.itemTextColor = itemTextColor;
    }

    public void setDisableItemTextColor(@ColorInt int disableItemTextColor) {
        this.disableItemTextColor = disableItemTextColor;
    }

    public void setItemTextAppearance(@StyleRes int itemTextAppearance) {
        this.itemTextAppearance = itemTextAppearance;
    }

    public void setDisableRadioButtonTint(@ColorInt int disableRadioButtonTint) {
        this.disableRadioButtonTint = disableRadioButtonTint;
    }

    public void setRadioButtonTint(@ColorInt int radioButtonTint) {
        this.radioButtonTint = radioButtonTint;
    }

    @Override
    public int getItemCount() {
        return roles.size();
    }

    public String getSelectedRole() {
        if (selectedPosition != -1) {
            return roles.get(selectedPosition);
        }
        return null;
    }

    public void setGroupData(List<String> roles, @Nullable GroupMember member, Group group) {
        if (roles != null && member != null && group != null) {
            this.group = group;
            this.roles = roles;
            if (member.getScope().equalsIgnoreCase(CometChatConstants.SCOPE_ADMIN)) {
                selectedPosition = 0;
            } else if (member.getScope().equalsIgnoreCase(CometChatConstants.SCOPE_MODERATOR)) {
                selectedPosition = 1;
            } else {
                selectedPosition = 2;
            }
            initialPosition = selectedPosition;
            notifyDataSetChanged();
            notifySelectionChangeListener();
        }
    }

    public void setOnScopeSelectionChangeListener(OnScopeSelectionChangeListener listener) {
        this.onScopeSelectionChangeListener = listener;
    }

    private void notifySelectionChangeListener() {
        if (onScopeSelectionChangeListener != null) {
            onScopeSelectionChangeListener.onSelectionChanged(selectedPosition == initialPosition);
        }
    }

    public interface OnScopeSelectionChangeListener {
        void onSelectionChanged(boolean isSameAsOriginal);
    }

    public static class RoleViewHolder extends RecyclerView.ViewHolder {
        TextView textViewRole;
        AppCompatRadioButton radioButton;

        public RoleViewHolder(@NonNull View itemView) {
            super(itemView);
            textViewRole = itemView.findViewById(R.id.textViewRole);
            radioButton = itemView.findViewById(R.id.radioButton);
        }
    }
}

File: chatuikit-kotlin/src/main/java/com/cometchat/uikit/kotlin/presentation/groupmembers/scopechange/ScopeAdapter.kt
package com.cometchat.uikit.kotlin.presentation.groupmembers.scopechange

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.StyleRes
import androidx.appcompat.widget.AppCompatRadioButton
import androidx.recyclerview.widget.RecyclerView
import com.cometchat.chat.constants.CometChatConstants
import com.cometchat.chat.models.Group
import com.cometchat.chat.models.GroupMember
import com.cometchat.uikit.kotlin.R

/**
 * Adapter for displaying scope/role options with radio buttons.
 * Mirrors the original Java ScopeAdapter exactly.
 *
 * Key behaviors:
 * - Moderators cannot assign Admin scope (radio button disabled, text dimmed)
 * - Current member scope is pre-selected
 * - Single selection via radio buttons
 */
class ScopeAdapter(
    private val context: Context,
    private var roles: List<String>?
) : RecyclerView.Adapter<ScopeAdapter.RoleViewHolder>() {

    private var selectedPosition = -1
    private var group: Group? = null

    @ColorInt private var itemTextColor: Int = 0
    @ColorInt private var disableItemTextColor: Int = 0
    @StyleRes private var itemTextAppearance: Int = 0
    @ColorInt private var disableRadioButtonTint: Int = 0
    @ColorInt private var radioButtonTint: Int = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RoleViewHolder {
        val view = LayoutInflater.from(context)
            .inflate(R.layout.cometchat_list_scope_change_item_radio, parent, false)
        return RoleViewHolder(view)
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onBindViewHolder(holder: RoleViewHolder, @SuppressLint("RecyclerView") position: Int) {
        val role = roles?.get(position) ?: return
        holder.textViewRole.text = role
        holder.radioButton.isChecked = holder.adapterPosition == selectedPosition
        if (itemTextAppearance != 0) {
            holder.textViewRole.setTextAppearance(itemTextAppearance)
        }

        // Moderator cannot assign Admin scope — matches original Java ScopeAdapter
        if (CometChatConstants.SCOPE_MODERATOR.equals(group?.scope, ignoreCase = true) &&
            role.equals(CometChatConstants.SCOPE_ADMIN, ignoreCase = true)
        ) {
            holder.radioButton.isEnabled = false
            holder.radioButton.buttonTintList = ColorStateList.valueOf(disableRadioButtonTint)
            holder.textViewRole.setTextColor(disableItemTextColor)
        } else {
            holder.textViewRole.setTextColor(itemTextColor)
            holder.radioButton.buttonTintList = ColorStateList.valueOf(radioButtonTint)
            holder.radioButton.isEnabled = true
            holder.itemView.setOnClickListener {
                selectedPosition = holder.adapterPosition
                notifyDataSetChanged()
            }
            holder.radioButton.setOnClickListener {
                selectedPosition = holder.adapterPosition
                notifyDataSetChanged()
            }
        }
    }

    override fun getItemCount(): Int = roles?.size ?: 0

    fun getSelectedRole(): String? {
        return if (selectedPosition != -1) roles?.get(selectedPosition) else null
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setGroupData(roles: List<String>, member: GroupMember?, group: Group?) {
        if (roles != null && member != null && group != null) {
            this.group = group
            this.roles = roles
            selectedPosition = when {
                member.scope.equals(CometChatConstants.SCOPE_ADMIN, ignoreCase = true) -> 0
                member.scope.equals(CometChatConstants.SCOPE_MODERATOR, ignoreCase = true) -> 1
                else -> 2
            }
            notifyDataSetChanged()
        }
    }

    fun setItemTextColor(@ColorInt color: Int) { itemTextColor = color }
    fun setDisableItemTextColor(@ColorInt color: Int) { disableItemTextColor = color }
    fun setItemTextAppearance(@StyleRes appearance: Int) { itemTextAppearance = appearance }
    fun setRadioButtonTint(@ColorInt color: Int) { radioButtonTint = color }
    fun setDisableRadioButtonTint(@ColorInt color: Int) { disableRadioButtonTint = color }

    class RoleViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val textViewRole: TextView = itemView.findViewById(R.id.textViewRole)
        val radioButton: AppCompatRadioButton = itemView.findViewById(R.id.radioButton)
    }
}
